include/zephyr/toolchain/gcc.h

```diff
@@ -683,4 +683,44 @@ do {                                                                    \
 	_Pragma("GCC diagnostic pop")
 
 #endif /* !_LINKER */
+
+#define TOOLCHAIN_WARNING_ADDRESS_OF_PACKED_MEMBER "-Waddress-of-packed-member"
+#define TOOLCHAIN_WARNING_ARRAY_BOUNDS "-Warray-bounds"
+#define TOOLCHAIN_WARNING_ATTRIBUTES "-Wattributes"
+#define TOOLCHAIN_WARNING_DANGLING_POINTER "-Wdangling-pointer"
+#define TOOLCHAIN_WARNING_DELETE_NON_VIRTUAL_DTOR "-Wdelete-non-virtual-dtor"
+#define TOOLCHAIN_WARNING_EXTRA "-Wextra"
+#define TOOLCHAIN_WARNING_FORMAT_TRUNCATION "-Wformat-truncation"
+#define TOOLCHAIN_WARNING_INFINITE_RECURSION "-Winfinite-recursion"
+#define TOOLCHAIN_WARNING_INTEGER_OVERFLOW "-Winteger-overflow"
+#define TOOLCHAIN_WARNING_NONNULL "-Wnonnull"
+#define TOOLCHAIN_WARNING_OVERFLOW "-Woverflow"
+#define TOOLCHAIN_WARNING_PRAGMAS "-Wpragmas"
+#define TOOLCHAIN_WARNING_SHADOW "-Wshadow"
+#define TOOLCHAIN_WARNING_UNUSED_FUNCTION "-Wunused-function"
+#define TOOLCHAIN_WARNING_UNUSED_LABEL "-Wunused-label"
+#define TOOLCHAIN_WARNING_UNUSED_VARIABLE "-Wunused-variable"
```

Comment:
This is not just adjusting what we do in special cases, but seems to embark on a path where Zephyr will start to disable warnings in general i C-code.

Please keep those changes out of the PR so that we can have two PRs
- One which clean up existing usage.
- One PR which propose to start disabling warnings in C code instead of the compiler in general.
   This allows us to have a discussion in the arch wg.

Reply:
Note, most of those are localized to `tests/` for disabling, which I think is fine, but this PR exposes the principle to be more generally available and this not local to test code, for example: `-Wunused-variable`, `-Winteger-overflow`, `-Winfinite-recursion`, etc.

Reply:
> One which clean up existing usage.
> One PR which propose to start disabling warnings in C code instead of the compiler in general.

I split the two commits that were in this PR into PRs, which I think is what you're asking for. The new PR is https://github.com/zephyrproject-rtos/zephyr/pull/84944.

Note that I would also prefer for all of the `TOOLCHAIN_WARNING_*` defines to not be needed. Right now they are simply the list of what is already in the code via `#pragma` (see https://github.com/zephyrproject-rtos/zephyr/pull/84065).

As a followup, I'm trying to remove any of these that no longer appear to be necessary in https://github.com/zephyrproject-rtos/zephyr/pull/84117. Hopefully it means the final list of `TOOLCHAIN_WARNING_*` will be shorter.

Reply:
> I split the two commits that were in this PR into PRs, which I think is what you're asking for. The new PR is https://github.com/zephyrproject-rtos/zephyr/pull/84944.

:+1: 

> Note that I would also prefer for all of the TOOLCHAIN_WARNING_* defines to not be needed. Right now they are simply the list of what is already in the code via #pragma (see https://github.com/zephyrproject-rtos/zephyr/pull/84065).

Yes, I know.
The main thing I just happened to notice is some of those addresses what is in Zephyr main tree, but some are in the test code parts.

Testing code usually allows a bit more freedom / tweaks which we should try to avoid / minimize in real code.
So let's try to keep the parts required for testing purposes localized to testing.